### PR TITLE
Allow SSZ builder submissions (also SSZ+GZIP)

### LINKED ISSUE #209: feat: allow block builder submissions with ssz payload, and with json+gzip

### PR DESCRIPTION
**Pull request by metachris**
## 📝 Summary

Allows block builder submissions also in SSZ, besides JSON.

- Requires https://github.com/attestantio/go-builder-client/pull/10 which fixes SSZ encoding
- PR includes tests for JSON, SSZ, JSON+GZIP and SSZ+GZIP submitNewBlock payloads
- Fixes #209 

| **Encoding** | **Bytes** |
|--------------|-----------|
| JSON         | 704810    |
| SSZ          | 352239    |
| JSON+GZIP    | 207788    |
| SSZ+GZIP     | 195923    |

## Decoding performance

```
BenchmarkDecoding/json
BenchmarkDecoding/json-10                  18710             63206 ns/op            5712 B/op         75 allocs/op
BenchmarkDecoding/ssz
BenchmarkDecoding/ssz-10                 5619360               224.4 ns/op           185 B/op          0 allocs/op
```

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race`
* [x] `go mod tidy`
* [x] I have seen and agree to `CONTRIBUTING.md`

### CODE REVIEW

File: common/ssz_test.go [hunk 1]
```diff
@@ -23,9 +23,12 @@ func TestSSZBuilderSubmission(t *testing.T) {
 	ssz, err := depositData.MarshalSSZ()
 	require.NoError(t, err)
 
+	// err = os.WriteFile("/tmp/test.ssz", ssz, 0644)
+	// require.NoError(t, err)
+
```
Comment (jtraglia) [inline common/ssz_test.go:28]: Let's remove this commented out code.

**Review by jtraglia — DISMISSED** [review]
Just a few little nits, but LGTM 👍 I'm all for performance improvements.

File: common/ssz_test.go (hunk 1 above)
Comment (metachris) [reply common/ssz_test.go]: done

File: services/api/service_test.go [hunk 2]
```diff
@@ -79,7 +84,25 @@ func newTestBackend(t require.TestingT, numBeaconNodes int) *testBackend {
...
+		req.Header.Set(k, v)
+	}
+
+	// lfg
+	rr := httptest.NewRecorder()
+	be.relay.getRouter().ServeHTTP(rr, req)
+	return rr
+}
+
+func (be *testBackend) request(method, path string, payload any, headers map[string]string) *httptest.ResponseRecorder {
```
[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service_test.go]: done

File: common/ssz_test.go [hunk 3]
```diff
@@ -44,11 +47,37 @@ func TestSSZGetHeaderResponse(t *testing.T) {
 	ssz, err := payload.Capella.MarshalSSZ()
 	require.NoError(t, err)
 
+	// err = os.WriteFile("/tmp/test.ssz", ssz, 0644)
+	// require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (metachris) [reply common/ssz_test.go]: done

File: services/api/service.go [hunk 4]
```diff
@@ -1387,7 +1411,7 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		log.Warn("could not find slot duty")
 		api.RespondError(w, http.StatusBadRequest, "could not find slot duty")
 		return
-	} else if slotDuty.Entry.Message.FeeRecipient.String() != payload.ProposerFeeRecipient() {
+	} else if !strings.EqualFold(slotDuty.Entry.Message.FeeRecipient.String(), payload.ProposerFeeRecipient()) {
```
[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go:1414]: yeah, was needed to work with both JSON and SSZ :)

**Review by jtraglia — APPROVED** [review]
LGTM. Thanks 😄

File: services/api/service_test.go [hunk 5]
```diff
@@ -79,6 +84,24 @@ func newTestBackend(t require.TestingT, numBeaconNodes int) *testBackend {
...
+
+	req, err = http.NewRequest(method, path, bytes.NewReader(payload))
+	require.NoError(be.t, err)
+
+	// Set headers
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+
+	// lfg
```
Comment (michaelneuder) [inline services/api/service_test.go:99]: nit: lfg?

**Review by michaelneuder — COMMENTED** [review]
LGTM! ssz is a great!

File: services/api/service_test.go (hunk 5 above)
Comment (metachris) [reply services/api/service_test.go:99]: let's fucking go 🤠

File: services/api/service.go [hunk 6]
```diff
@@ -1299,23 +1300,47 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
 			api.RespondError(w, http.StatusBadRequest, err.Error())
 			return
 		}
-		log = log.WithField("gzip-req", true)
 	}
 
-	payload := new(common.BuilderSubmitBlockRequest)
-	if err := json.NewDecoder(r).Decode(payload); err != nil {
-		log.WithError(err).Warn("could not decode payload")
+	limitReader := io.LimitReader(r, 10*1024*1024) // 10 MB
```
[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go:1314]: that's correct! let's update sometime later.

**Pull request merged**